core_backend/app/dashboard/models.py

```diff
@@ -282,6 +286,148 @@ async def get_timeseries_urgency(
     return {row.time_period.strftime(format_str): row.n_urgent for row in rows}
 
 
+async def get_timeseries_top_content(
+    user_id: int,
+    asession: AsyncSession,
+    top_n: int,
+    start_date: date,
+    end_date: date,
+    frequency: TimeFrequency,
+) -> List[TopContentTimeSeries]:
+    """
+    Retrieve most frequently shared content.
+    Note that this retrieves top N content from the `QueryResponseContentDB` table
+    and not from the `ContentDB` table.ContentDB
+
+    Returns
+    """
+
+    interval_str, ts_labels = get_time_labels_query(frequency, start_date, end_date)
+
+    top_content = (
+        select(
+            ContentDB.content_id,
+            ContentDB.content_title,
+            func.count(QueryResponseContentDB.query_id).label("total_query_count"),
+        )
+        .select_from(QueryResponseContentDB)
+        .join(
+            ContentDB,
+            QueryResponseContentDB.content_id == ContentDB.content_id,
+        )
+        .where(
+            ContentDB.user_id == user_id,
+            QueryResponseContentDB.created_datetime_utc >= start_date,
+            QueryResponseContentDB.created_datetime_utc < end_date,
+        )
+        .group_by(
+            ContentDB.content_title,
+            ContentDB.content_id,
+        )
+        .order_by(desc("total_query_count"))
+        .limit(top_n)
+        .subquery()
+    )
+
+    all_combinations = (
+        select(
+            ts_labels.c.time_period,
+            top_content.c.content_id,
+            top_content.c.content_title,
+            top_content.c.total_query_count,
+        )
+        .select_from(ts_labels)
+        .join(top_content, text("1=1"))
```

Comment:
This is how you do an outer join in SQL Alchemy. Basically join on everything. Can also do "literal(True)" for the second argument.